File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/analyticssubmission/AnalyticsSubmissionService.java
package uk.nhs.nhsx.analyticssubmission;

import com.amazonaws.services.kinesisfirehose.AmazonKinesisFirehose;
import com.amazonaws.services.kinesisfirehose.model.PutRecordRequest;
import com.amazonaws.services.kinesisfirehose.model.Record;
import org.apache.http.entity.ContentType;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import uk.nhs.nhsx.analyticssubmission.model.ClientAnalyticsSubmissionPayload;
import uk.nhs.nhsx.analyticssubmission.model.StoredAnalyticsSubmissionPayload;
import uk.nhs.nhsx.core.Jackson;
import uk.nhs.nhsx.core.aws.s3.ByteArraySource;
import uk.nhs.nhsx.core.aws.s3.Locator;
import uk.nhs.nhsx.core.aws.s3.ObjectKeyNameProvider;
import uk.nhs.nhsx.core.aws.s3.S3Storage;
import uk.nhs.nhsx.core.events.Events;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

public class AnalyticsSubmissionService {

    private final static Logger log = LogManager.getLogger(AnalyticsSubmissionService.class);

    private final AnalyticsConfig config;
    private final S3Storage s3Storage;
    private final ObjectKeyNameProvider objectKeyNameProvider;
    private final AmazonKinesisFirehose kinesisFirehose;
    private final Events events;

    public AnalyticsSubmissionService(AnalyticsConfig config,
                                      S3Storage s3Storage,
                                      ObjectKeyNameProvider objectKeyNameProvider,
                                      AmazonKinesisFirehose kinesisFirehose,
                                      Events events) {
        this.config = config;
        this.s3Storage = s3Storage;
        this.objectKeyNameProvider = objectKeyNameProvider;
        this.kinesisFirehose = kinesisFirehose;
        this.events = events;
    }

    public void accept(ClientAnalyticsSubmissionPayload payload) {
        String json = Jackson.toJson(StoredAnalyticsSubmissionPayload.convertFrom(payload, events));

        if (config.s3IngestEnabled) {
            uploadToS3(json);
        }

        if (config.firehoseIngestEnabled) {
            uploadToFirehose(json);
        }
    }


    private void uploadToS3(String json) {
        var objectKey = objectKeyNameProvider.generateObjectKeyName().append(".json");

        s3Storage.upload(
            Locator.of(config.bucketName, objectKey),
            ContentType.APPLICATION_JSON,
            ByteArraySource.fromUtf8String(json)
        );
    }

    private void uploadToFirehose(String json) {
        var data = ByteBuffer.wrap(json.getBytes(StandardCharsets.UTF_8));
        var record = new Record().withData(data);

        var putRecordRequest = new PutRecordRequest()
                .withRecord(record)
                .withDeliveryStreamName(config.firehoseStreamName);

        log.info("Sending json to {}", config.firehoseStreamName);

        kinesisFirehose.putRecord(putRecordRequest);
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/analyticssubmission/AnalyticsSubmissionService.kt
package uk.nhs.nhsx.analyticssubmission

import com.amazonaws.services.kinesisfirehose.AmazonKinesisFirehose
import com.amazonaws.services.kinesisfirehose.model.PutRecordRequest
import com.amazonaws.services.kinesisfirehose.model.Record
import org.apache.http.entity.ContentType
import uk.nhs.nhsx.analyticssubmission.UploadType.Firehose
import uk.nhs.nhsx.analyticssubmission.UploadType.S3
import uk.nhs.nhsx.analyticssubmission.model.ClientAnalyticsSubmissionPayload
import uk.nhs.nhsx.analyticssubmission.model.StoredAnalyticsSubmissionPayload
import uk.nhs.nhsx.core.Jackson.toJson
import uk.nhs.nhsx.core.aws.s3.ByteArraySource.Companion.fromUtf8String
import uk.nhs.nhsx.core.aws.s3.Locator
import uk.nhs.nhsx.core.aws.s3.ObjectKey
import uk.nhs.nhsx.core.aws.s3.ObjectKeyNameProvider
import uk.nhs.nhsx.core.aws.s3.S3Storage
import uk.nhs.nhsx.core.events.Events
import java.nio.ByteBuffer

class AnalyticsSubmissionService(
    private val config: AnalyticsConfig,
    private val s3Storage: S3Storage,
    private val objectKeyNameProvider: ObjectKeyNameProvider,
    private val kinesisFirehose: AmazonKinesisFirehose,
    private val events: Events
) {
    fun accept(payload: ClientAnalyticsSubmissionPayload?) {
        val json = toJson(StoredAnalyticsSubmissionPayload.convertFrom(payload, events))
        val objectKey = objectKeyNameProvider.generateObjectKeyName()

        if (config.s3IngestEnabled) {
            uploadToS3(json, objectKey)
        }

        if (config.firehoseIngestEnabled) {
            uploadToFirehose(json, objectKey)
        }
    }

    private fun uploadToS3(json: String, objectKey: ObjectKey) {
        s3Storage.upload(
            Locator.of(config.bucketName, objectKey.append(".json")),
            ContentType.APPLICATION_JSON,
            fromUtf8String(json)
        )

        events(AnalyticsSubmissionUploaded::class.java, AnalyticsSubmissionUploaded(S3, config.firehoseStreamName, objectKey))
    }

    private fun uploadToFirehose(json: String, objectKey: ObjectKey) {
        val record = Record().withData(ByteBuffer.wrap(json.toByteArray()))
        val putRecordRequest = PutRecordRequest()
            .withRecord(record)
            .withDeliveryStreamName(config.firehoseStreamName)
        kinesisFirehose.putRecord(putRecordRequest)

        events(AnalyticsSubmissionUploaded::class.java, AnalyticsSubmissionUploaded(Firehose, config.firehoseStreamName, objectKey))
    }
}
